File: MurderMystery.hs
import Data.List (permutations)

-- Типи даних
data Person = George | John | Robert | Barbara | Christine | Yolanda
  deriving (Eq, Show, Enum, Bounded)

data Room = Bathroom | DiningRoom | Kitchen | LivingRoom | Pantry | Study
  deriving (Eq, Show, Enum, Bounded)

data Weapon = Bag | Firearm | Gas | Knife | Poison | Rope
  deriving (Eq, Show, Enum, Bounded)

type Assignment = [(Person, Room, Weapon)]

people :: [Person]
people = [minBound .. maxBound]

rooms :: [Room]
rooms = [minBound .. maxBound]

weapons :: [Weapon]
weapons = [minBound .. maxBound]

-- Генерація всіх можливих комбінацій
assignments :: [Assignment]
assignments =
  [ zip3 people rs ws
  | rs <- permutations rooms
  , ws <- permutations weapons
  ]

-- Допоміжні функції
roomOf :: Person -> Assignment -> Room
roomOf p a = head [r | (p', r, _) <- a, p' == p]

weaponOf :: Person -> Assignment -> Weapon
weaponOf p a = head [w | (p', _, w) <- a, p' == p]

personInRoom :: Room -> Assignment -> Person
personInRoom r a = head [p | (p, r', _) <- a, r' == r]

weaponInRoom :: Room -> Assignment -> Weapon
weaponInRoom r a = head [w | (_, r', w) <- a, r' == r]

-- Підказки
valid :: Assignment -> Bool
valid a =
  and
    [ -- Clue 1
      let p = personInRoom Kitchen a
          w = weaponInRoom Kitchen a
       in p `elem` [George, John, Robert]
          && w `notElem` [Rope, Knife, Bag, Firearm]

    , -- Clue 2
      (roomOf Barbara a == Study && roomOf Yolanda a == Bathroom)
      || (roomOf Barbara a == Bathroom && roomOf Yolanda a == Study)

    , -- Clue 3
      let bagOwner = head [p | (p, _, w) <- a, w == Bag]
          bagRoom  = roomOf bagOwner a
       in bagOwner `notElem` [Barbara, George]
          && bagRoom `notElem` [Bathroom, DiningRoom]

    , -- Clue 4
      let ropeOwner = head [p | (p, _, w) <- a, w == Rope]
       in ropeOwner `elem` [Barbara, Christine, Yolanda]
          && roomOf ropeOwner a == Study

    , -- Clue 5
      let p = personInRoom LivingRoom a
       in p `elem` [John, George]

    , -- Clue 6
      weaponInRoom DiningRoom a /= Knife

    , -- Clue 7
      let w = weaponOf Yolanda a
       in w /= weaponInRoom Study a
          && w /= weaponInRoom Pantry a

    , -- Clue 8
      weaponOf George a == Firearm

    , -- Final clue
      weaponInRoom Pantry a == Gas
    ]

solutions :: [Assignment]
solutions = filter valid assignments

main :: IO ()
main = do
  let solution = head solutions
      murderer = personInRoom Pantry solution
  putStrLn "Solution:"
  mapM_ print solution
  putStrLn ("\nMurderer: " ++ show murderer)

File: MurderMystery.pl
:- use_module(library(lists)).

% головний предикат
solve(Murderer) :-
    People = [george, john, robert, barbara, christine, yolanda],
    Rooms  = [bathroom, dining_room, kitchen, living_room, pantry, study],
    Weapons = [bag, firearm, gas, knife, poison, rope],

    % представлення: person(Person, Room, Weapon)
    Solution = [
        person(george, RG, WG),
        person(john, RJ, WJ),
        person(robert, RR, WR),
        person(barbara, RB, WB),
        person(christine, RC, WC),
        person(yolanda, RY, WY)
    ],

    % всі кімнати різні
    permutation(Rooms, [RG, RJ, RR, RB, RC, RY]),
    % всі знаряддя різні
    permutation(Weapons, [WG, WJ, WR, WB, WC, WY]),

    % ---------- Clue 1 ----------
    % чоловік на кухні не мав rope, knife, bag і не firearm
    member(person(P1, kitchen, WK), Solution),
    member(P1, [george, john, robert]),
    \+ member(WK, [rope, knife, bag, firearm]),

    % ---------- Clue 2 ----------
    % Barbara і Yolanda: study / bathroom
    member(person(barbara, RB, _), Solution),
    member(person(yolanda, RY, _), Solution),
    member(RB, [study, bathroom]),
    member(RY, [study, bathroom]),
    RB \= RY,

    % ---------- Clue 3 ----------
    % bag не у Barbara і не у George
    member(person(PB, RBAG, bag), Solution),
    PB \= barbara,
    PB \= george,
    RBAG \= bathroom,
    RBAG \= dining_room,

    % ---------- Clue 4 ----------
    % жінка з rope була в study
    member(person(PR, study, rope), Solution),
    member(PR, [barbara, christine, yolanda]),

    % ---------- Clue 5 ----------
    % у living_room був John або George
    member(person(P5, living_room, _), Solution),
    member(P5, [john, george]),

    % ---------- Clue 6 ----------
    % knife не в dining_room
    member(person(_, RK, knife), Solution),
    RK \= dining_room,

    % ---------- Clue 7 ----------
    % Yolanda не мала зброї зі study і pantry
    member(person(_, study, WS), Solution),
    member(person(_, pantry, WP), Solution),
    member(person(yolanda, _, WY), Solution),
    WY \= WS,
    WY \= WP,

    % ---------- Clue 8 ----------
    % firearm з George
    member(person(george, RG, firearm), Solution),

    % ---------- Final ----------
    % газ у pantry → вбивця
    member(person(Murderer, pantry, gas), Solution).
